src/Datadog.Trace.DuckTyping/README.md

```diff
@@ -0,0 +1,757 @@
+}
+```
+
+This attribute can be used in `Properties` or `Methods` in the proxy definition and is used by the library to select the members we want to access.
+
+### Example
+
+The following example shows multiple uses of the `DuckAttribute`:
+
+```csharp
+public interface IMyProxy
+{
+    // ***
+    // *** Field binding
+    // ***
+
+    [Duck(Name = "_sampleStaticField", Kind = DuckKind.Field)]
+    string MyStaticField { get; }
+
+    [Duck(Name = "_normalField", Kind = DuckKind.Field)]
+    int NormalFieldWithGetterAndSetter { get; set; }
+
+
+
+    // ***
+    // *** Property binding (by default the Kind is Property so we can ignore it)
+    // ***
+
+    [Duck(Name = "Instance")]
+    string MyStaticInstanceProperty { get; }
+
+    // If we don´t use the duck attribute it will try to find a property with the same name in the target type
+    string MyProperty { get; set; }
+
+    // If the original property returns a type that we don´t have at compile time,
+    // we can pass a proxy definition and that original value will be wrapped with the proxy
+    // as well (DuckType chaining).
+    IObscureObject CustomData { get; }
+
+    // Indexers are supported as well
+    string this[string index] { get; set; }
+
+
+
+    // ***
+    // *** Method binding
+    // ***
+
+    // support overloads of a method
+    void Add(string name, int obj);
+    void Add(string name, string obj = "none");
+
+    // In order to resolve an ambiguity of a call we can specify the parameters we want to match.
+    [Duck(ParameterTypeNames = new string[] { "System.String", "MyNamespace.MyType, MyAssembly" })]
+    void Add(string name, object obj);
+
+    // We can also use the name to map to another method name.
+    [Duck(Name = "InternalGetReference")]
+    void GetReference(ref int value);
+
+    // The Proxy definition can support generics only if the instance type
+    // is *Public* 
+    Tuple<T1, T2> Wrap<T1, T2>(T1 a, T2 b);
+
+    // If we need to call a generic method from a non public type
+    // we can use the DuckAttribute as well:
+    [Duck(Name = "Wrap", GenericParameterTypeNames = new[] { "System.Int32", "System.String" })]
+    Tuple<int, string> WrapIntString(int a, string b);
+
+
+    // If the original call returns or uses a parameter that we don´t have at compile time,
+    // we can pass a proxy definition and that original value will be wrapped with the proxy
+    // as well (DuckType chaining).
+    bool TryGetPrivateObscure(out IObscureObject obj);
+}
+```
+
+
+## Accessor modifiers (AM)
+
+In order to support all accessor modifiers for: instance types, parameters and return value types, the Duck Type library applies some `tricks` to avoid the visibility checks. This is done automatically when the library is creating the proxy type. In summary the following logic is applied depending on each case:
+
+|      Target Instance AM      | Target Member Type |       Target Member AM       | Access Method                                 |
```

Comment:
Instead of "Target Instance AM" could we use "Target Type AM"?

Reply:
Sure!, done in https://github.com/DataDog/dd-trace-dotnet/pull/946/commits/564fcdf77e48b4645b9881962e5bcd6dd38ee9f5